hypershift-operator/controllers/nodepool/haproxy.go

```diff
@@ -0,0 +1,377 @@
+
+	"github.com/clarketm/json"
+	ignitionapi "github.com/coreos/ignition/v2/config/v3_2/types"
+	api "github.com/openshift/hypershift/api"
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/control-plane-operator/controllers/hostedcontrolplane/ignition"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/hostedcluster"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/manifests"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/manifests/controlplaneoperator"
+	"github.com/openshift/hypershift/support/config"
+	"github.com/openshift/hypershift/support/globalconfig"
+	"github.com/openshift/hypershift/support/releaseinfo"
+	"github.com/openshift/hypershift/support/util"
+	mcfgv1 "github.com/openshift/hypershift/thirdparty/machineconfigoperator/pkg/apis/machineconfiguration.openshift.io/v1"
+	"github.com/vincent-petithory/dataurl"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/utils/pointer"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	controlPlaneOperatorSkipsHAProxyConfigGenerationLabel = "io.openshift.hypershift.control-plane-operator-skips-haproxy"
+)
+
+func (r *NodePoolReconciler) reconcileHAProxyIgnitionConfig(ctx context.Context, releaseImage *releaseinfo.ReleaseImage, nodepool *hyperv1.NodePool, hcluster *hyperv1.HostedCluster, globalConfig globalconfig.GlobalConfig) error {
+	var pullSecret corev1.Secret
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: hcluster.Namespace, Name: hcluster.Spec.PullSecret.Name}, &pullSecret); err != nil {
+		return fmt.Errorf("failed to get pull secret: %w", err)
+	}
+	pullSecretBytes, ok := pullSecret.Data[corev1.DockerConfigJsonKey]
+	if !ok {
+		return fmt.Errorf("expected %s key in pull secret", corev1.DockerConfigJsonKey)
+	}
+	controlPlaneOperatorImage, err := hostedcluster.GetControlPlaneOperatorImage(ctx, hcluster, r.ReleaseProvider, r.HypershiftOperatorImage, pullSecretBytes)
+	if err != nil {
+		return fmt.Errorf("failed to get controlPlaneOperatorImage: %w", err)
+	}
+	controlPlaneOperatorImageMetadata, err := r.ImageMetadataProvider.ImageMetadata(ctx, controlPlaneOperatorImage, pullSecretBytes)
+	if err != nil {
+		return fmt.Errorf("failed to look up image metadata for %s: %w", controlPlaneOperatorImage, err)
+	}
+	// Don't manage the configmap or we will race with the CPO
+	if _, cpoSkips := util.ImageLabels(controlPlaneOperatorImageMetadata)[controlPlaneOperatorSkipsHAProxyConfigGenerationLabel]; !cpoSkips {
+		return nil
+	}
+
+	hcpNS := manifests.HostedControlPlaneNamespace(hcluster.Namespace, hcluster.Name).Name
+	hcp := controlplaneoperator.HostedControlPlane(hcpNS, hcluster.Name)
+	if err := r.Client.Get(ctx, crclient.ObjectKeyFromObject(hcp), hcp); err != nil {
```

Comment:
That would mean a) an api change b) just move the problem by introducing this dep into the hostedcluster controller instead which c) means we have that dep there and it is not apparent there what it is even needed for.

Reply:
I'm not seeing why a, b and c? what I mean is that the HC creates and dictates the input for the HCP and the HCP signals back to the HC which reflects state of the world in status/conditions. NodePools are not even aware of a HCP even exist atm. All the info we are getting here we should be able to get it from the HC spec and conditions. Why do we need to inspect the HCP? If we really need to may be this is surfacing a design flaw that indeed requires some API signalling and defaulting polishing to avoid accessing the HCP directly. I think having the NodePool controller to look at the HCP is crossing a design boundary and increases complexity.

Reply:
We need to inspect the HCP to find out the KAS endpoint address and port. We do not propagate this up today, what we do propgate up is the kubeconfig reference because there is no known case where consumers would want the endpoint address but not the credentials to actually be able to use it.

Propagating this up means:
* We introduce a field in our external API object that we do not actually want to expose externally just because we need it internally
* We need to add a new feature flag to tell the HO if that field exists (and find a workaround like loading the kubeconfig if it doesn't)
* We are making the dependency of "Nodepool controller needs to know api server address and port" harder to understand by introducing a new layer of indirection (hostedcluster controller copies from internal api to external api for internal consumer only)

> I think having the NodePool controller to look at the HCP is crossing a design boundary and increases complexity.

IMHO setting such a boundary is pretty arbitrary, both the nodepool controller and the HCP are part of the project and their existence is an implementation detail. Propagating it up is significantly more complex, adds a field to our external api that we will never be able to remove and don't actually want there and makes it harder to understand what it is needed for - While it is true that it is not ideal to access the HCP, it is IMHO the by far better of the two trade-offs we have to choose from here.

Reply:
@alvaroaleman Thanks for elaborating.
The input the nodePool controller is using (apiServerInternalAddress and apiServerInternalPort) is coming form the HC and dictated by the HC, which then passes it to the HCP via [`reconcileHostedControlPlane`](https://github.com/openshift/hypershift/blob/7652986cffc8a633342410ea164c4ee94a3b9b62/hypershift-operator/controllers/hostedcluster/hostedcluster_controller.go#L1222-L1226):

```
	if hcluster.Spec.Networking.APIServer != nil {
		hcp.Spec.APIAdvertiseAddress = hcluster.Spec.Networking.APIServer.AdvertiseAddress
		hcp.Spec.APIPort = hcluster.Spec.Networking.APIServer.Port
	}
```

In line 84 this PR is doing:
```
	if hcp.Spec.APIAdvertiseAddress != nil {
		apiServerInternalAddress = *hcp.Spec.APIAdvertiseAddress
	} else {
		apiServerInternalAddress = config.DefaultAdvertiseAddress
	}
	if hcp.Spec.APIPort != nil {
		apiServerInternalPort = *hcp.Spec.APIPort
	} else {
		apiServerInternalPort = config.DefaultAPIServerPort
	}
```

Is there a reason why can't we instead do as below? Why is fetching the HCP needed?:
```
	if hcluster.Spec.Networking.APIServer.AdvertiseAddress != nil {
		apiServerInternalAddress = *hcluster.Spec.Networking.APIServer.AdvertiseAddress
	} else {
		apiServerInternalAddress = config.DefaultAdvertiseAddress
	}
	if hcluster.Spec.Networking.APIServer.Port != nil {
		apiServerInternalPort = *hcluster.Spec.Networking.APIServer.Port
	} else {
		apiServerInternalPort = config.DefaultAPIServerPort
	}
```